ffs/scheduler/internal/sjstore/sjstore.go

```diff
@@ -280,12 +298,38 @@ func (s *Store) put(j ffs.StorageJob) error {
 	if err := s.ds.Put(makeKey(j.ID), buf); err != nil {
 		return fmt.Errorf("saving to datastore: %s", err)
 	}
-	s.notifyWatchers(j)
+
+	// Update executing cids cache.
 	if j.Status == ffs.Executing {
 		s.executingCids[j.Cid] = j.ID
-	} else if j.Status == ffs.Failed || j.Status == ffs.Success || j.Status == ffs.Canceled {
-		delete(s.executingCids, j.Cid)
+	} else {
+		execJobID, ok := s.executingCids[j.Cid]
+		// If the executing job is not longer executing,
+		// take out from executing cache.
+		if ok && execJobID == j.ID {
+			delete(s.executingCids, j.Cid)
+		}
```

Comment:
If the updated Job isn't executing, no matter which this state is, shouldn't be in the executing cache.